CONTRIBUTING.md

```diff
@@ -0,0 +1,408 @@
+
+#### Naming style
+All Ginkgo specific variables should be prefixed with a `GINKGO_` and all functions by `ginkgo_`.
+
+
+## Helper scripts
+
+To facilitate easy development within Ginkgo and to encourage coders and scientists who do not want get bogged down by the details of the Ginkgo library, but rather focus on writing the algorithms and the kernels, Ginkgo provides the developers with a few helper scripts.
+
+### Create a new algorithm
+
+A `create_new_algorithm.sh` script is available for developers to facilitate easy addition of new algorithms. The options it provides can be queried with
+
+```sh
+./create_new_algorithm.sh --help
+```
+The main objective of this script is to add files and boiler plate code for the new algorithm using a model and an instance of that model. For example, models can be any one of`factorization`,`matrix`,`preconditioner` or `solver`. For example to create a new solver named `my_solver` similar to `gmres`, you would set the `ModelType` to `solver` and set the `ModelName` to `gmres`. This would duplicate the core algorithm and kernels of the `gmres` algorithm and replace the naming to `my_solver`. Additionally, marks all the kernels of the new `my_solver` to `GKO_NOT_IMPLEMENTED`.
+
+### Converting CUDA code to HIP code
+
+This script calls the `hipify` script provided by HIP converting the CUDA syntax to HIP syntax. Additionally, it also automatically replaces the instances of CUDA with HIP as appropriate. Hence this script can be called on a Ginkgo CUDA file. 
+
+
+## Writing Tests
+
+Ginkgo uses the [GTest framework](https://github.com/google/googletest) for the unit test framework within Ginkgo. Writing good tests are extremely important to verify the functionality of the new code and to make sure that none of the existing code has been broken.
+
+### Testing know-how
+
+* GTest provides a [comprehensive documentation](https://github.com/google/googletest/blob/master/googletest/docs/primer.md) of the functionality available within Gtest.  
+* Reduce code duplication with [Testing Fixtures, `TEST_F`](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests)
+* Write templated tests using [`TYPED_TEST`](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#typed-tests).
+
+### Some general rules.
+
+* Unit tests must follow the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle).
+* Unit tests must follow the [AAA](http://wiki.c2.com/?ArrangeActAssert) pattern, and a single blank line must appear between consecutive "A" sections.
+
+### Writing tests for kernels
+
+* Reference kernels, kernels on the `ReferenceExecutor`, are meant to be single threaded reference implementations. Therefore, tests for reference kernels need to be performed with data that can be as small as possible. For example, matrices lesser than 5x5 are acceptable. 
```

Comment:
It's also because we want exact testing, e.g. replicate the results of Matlab for example or something like that. Doing so with exact values with big matrices would be tedious.